vault/identity_store.go

```diff
@@ -621,145 +623,159 @@ func (i *IdentityStore) Invalidate(ctx context.Context, key string) {
-			entity, err := i.parseEntityFromBucketItem(ctx, item)
+			bucketEntity, err := i.parseEntityFromBucketItem(ctx, item)
 			if err != nil {
 				i.logger.Error("failed to parse entity from bucket entry item", "error", err)
 				return
 			}
 
-			localAliases, err := i.parseLocalAliases(entity.ID)
+			localAliases, err := i.parseLocalAliases(bucketEntity.ID)
 			if err != nil {
 				i.logger.Error("failed to load local aliases from storage", "error", err)
 				return
 			}
+
 			if localAliases != nil {
 				for _, alias := range localAliases.Aliases {
-					entity.UpsertAlias(alias)
+					bucketEntity.UpsertAlias(alias)
 				}
 			}
 
-			// Only update MemDB and don't touch the storage
-			err = i.upsertEntityInTxn(ctx, txn, entity, nil, false)
-			if err != nil {
-				i.logger.Error("failed to update entity in MemDB", "error", err)
-				return
+			var memDBEntity *identity.Entity
+			for i, entity := range memDBEntities {
+				if entity.ID == bucketEntity.ID {
+					memDBEntity = entity
+
+					// Remove this processed entity from the slice, so that
+					// all tht will be left are unprocessed entities.
+					copy(memDBEntities[i:], memDBEntities[i+1:])
+					memDBEntities = memDBEntities[:len(memDBEntities)-1]
+					break
+				}
 			}
 
-			// If we are a secondary, the entity created by the secondary
-			// via the CreateEntity RPC would have been cached. Now that the
-			// invalidation of the same has hit, there is no need of the
-			// cache. Clearing the cache. Writing to storage can't be
-			// performed by perf standbys. So only doing this in the active
-			// node of the secondary.
-			if i.localNode.ReplicationState().HasState(consts.ReplicationPerformanceSecondary) && i.localNode.HAState() == consts.Active {
-				if err := i.localAliasPacker.DeleteItem(ctx, entity.ID+tmpSuffix); err != nil {
-					i.logger.Error("failed to clear local alias entity cache", "error", err, "entity_id", entity.ID)
+			// If the entity is not in MemDB or if it is but differs from the
+			// state that's in the bucket storage entry, upsert it into MemDB.
+			if memDBEntity == nil || !reflect.DeepEqual(memDBEntity, bucketEntity) {
+				// The entity is not in MemDB, it's a new entity. Add it to MemDB.
+				err = i.upsertEntityInTxn(ctx, txn, bucketEntity, nil, false)
+				if err != nil {
+					i.logger.Error("failed to update entity in MemDB", "entity_id", bucketEntity.ID, "error", err)
 					return
 				}
-			}
 
-			entityIDs = append(entityIDs, entity.ID)
+				// If this is a performance secondary, the entity created on
+				// this node would have been cached in a local cache based on
+				// the result of the CreateEntity RPC call to the primary
+				// cluster. Since this invalidation is signaling that the
+				// entity is now in the primary cluster's storage, the locally
+				// cached entry can be removed.
+				if i.localNode.ReplicationState().HasState(consts.ReplicationPerformanceSecondary) && i.localNode.HAState() != consts.PerfStandby {
```

Comment:
Can we look into why this changed and make sure we're confident in it?

Reply:
Yes, that change is coming.